src/sas/qtgui/Utilities/DocViewWidget.py

```diff
@@ -43,19 +45,41 @@ def __init__(self,
         self.updatefn = updatefn
         self.reset_flag = reset_flag
         self.target = Path(target)
+        self.runner = None
+        self.documentation_lock = documentation_lock
+        self.communicate = Communicate()
+        self.communicate.closeSignal.connect(self.close)
 
     def compute(self):
         """
         Regen the docs in a separate thread
         """
         try:
-            if self.target.exists():
-                make_documentation(self.target)
+            if self.documentation_lock.acquire(blocking=False):
+                try:
+                    # Start a try/finally block to ensure that the lock is released if an exception is thrown
+                    if self.target.exists():
+                        self.runner = make_documentation(self.target)
+                        while self.runner.poll() is None:
+                            time.sleep(0.5)
+                            self.communicate.documentationUpdateLogSignal.emit()
+                finally:
+                    self.documentation_lock.release()
+                    self.runner.kill()
+                    self.runner = None
             else:
-                return
+                logging.info('Documentation regeneration is already in process. '
+                            'Please wait for it to complete before attempting again.')
         except KeyboardInterrupt as msg:
+            self.close()
             logging.log(0, msg)
 
+    def close(self):
+        # Ensure the runner and locks are fully released when closing the main application
+        if self.runner:
```

Comment:
I wonder what happens if the `closeSignal` from line 51 is triggered while in the polling loop of of line 63? It looks like  `self.documentation_lock.release(); self.runner.kill` will be called on line 67-68 will be called. Multiple `release` should be harmless. Not sure about multiple `kill`. Depending on the exact timing of `closeSignal` and another call to `compute`, the new runner might itself be killed.

Separate question: do you need some sort of `wait` on the killed thread?

Third question: is `closeSignal` triggered from the same thread as `compute` is running? Does `close` even have access to the runner process handles?

I'm not familiar enough with Qt signals and threads, so maybe the code is fine.

It may be easier to work with a (semi-)permanent documentation thread, and a queue of documentation build requests. You still may want a separate process within that thread for cancelling in-progress builds and capturing failed builds.  It'll be be easier to reason about with a clear separation from Qt signals.

Alternatively, use only Qt multiprocessing and signaling for the doc builds, assuming they have tools for managing  worker queues within their environment.